sklearn/manifold/_barnes_hut_tsne.pyx

```diff
@@ -72,25 +75,24 @@ cdef float compute_gradient(float[:] val_P,
     cdef float* neg_f = <float*> malloc(sizeof(float) * n_samples * n_dimensions)
     cdef float* pos_f = <float*> malloc(sizeof(float) * n_samples * n_dimensions)
 
-    sum_Q[0] = 0.0
     if take_timing:
         t1 = clock()
-    compute_gradient_negative(pos_reference, neg_f, qt, sum_Q,
-                              dof, theta, start, stop)
+    sQ = compute_gradient_negative(pos_reference, neg_f, qt, dof, theta, start,
+                                   stop, num_threads)
     if take_timing:
         t2 = clock()
         printf("[t-SNE] Computing negative gradient: %e ticks\n", ((float) (t2 - t1)))
-    sQ = sum_Q[0]
 
     if take_timing:
         t1 = clock()
     error = compute_gradient_positive(val_P, pos_reference, neighbors, indptr,
                                       pos_f, n_dimensions, dof, sQ, start,
-                                      qt.verbose, compute_error)
+                                      qt.verbose, compute_error, num_threads)
     if take_timing:
         t2 = clock()
-        printf("[t-SNE] Computing positive gradient: %e ticks\n", ((float) (t2 - t1)))
-    for i in range(start, n_samples):
+        printf("[t-SNE] Computing positive gradient: %e ticks\n",
+               ((float) (t2 - t1)))
+    for i in prange(start, n_samples, nogil=True, num_threads=num_threads):
```

Comment:
Maybe explicitly set `schedule='static'`, same elsewhere

Reply:
Why should we explicitly set the schedule ? static is the default

Reply:
> explicitly set the schedule ?

Because the default schedule is implementation-dependent as far as I understand (see link below)

And even if it were the default schedule, explicitly setting it gives a soft hint for people reading the code. It indicates how the code is parallelized.

> static is the default

Are you sure? The Cython doc doesn't mention any default, and https://stackoverflow.com/questions/21291581/openmp-parallel-for-what-is-default-schedule

Reply:
I have no opinion here, let me know what you think is the best.

Reply:
You're right. Actually the default of gnu implem is 'dynamic' :)

Reply:
for llvm and intel, the default is 'static'.

Reply:
however, setting the schedule prevents a user to set OMP_SCHEDULE.

Reply:
I think it's a good thing: you really want the schedule to be static here. Relying on `OMP_SCHEDULE` would make it possible to spawn one thread per sample, which would be catastrophic.
I'm not an expert but typically in the LightGBM code, they always hardcode the scheduling in the `#pragma omp...` loops

Reply:
yes it's probably the best here since we expect each step of the loop to take approximately the same time.

> Relying on OMP_SCHEDULE would make it possible to spawn one thread per sample

How can you do that ? as far as I know openmp spawns the threadpool once and then dispatch the tasks to the threads based on the schedule.

Reply:
What I mean is that ff the schedule is dynamic the default chunksize is 1. So with `for i in prange(n_samples)` you would have one thread per sample.